tests/unit/test_logwriter.c

```diff
@@ -83,72 +156,46 @@ testcase(gchar *msg_string, gboolean input_is_rfc5424, gchar *template, guint wr
   opt.options = LWO_NO_MULTI_LINE | LWO_NO_STATS | LWO_SHARE_STATS;
   opt.template_options.time_zone_info[LTZ_SEND]=tzinfo;
 
-  if (template)
+  if (c.template)
     {
       templ = log_template_new(configuration, "dummy");
-      log_template_compile(templ, template, &error);
+      log_template_compile(templ, c.template, &error);
     }
   opt.template = templ;
-  msg = init_msg(msg_string, input_is_rfc5424);
+  msg = init_msg(c.msg, c.is_rfc5424);
   queue = log_queue_fifo_new(1000, NULL);
-  writer = log_writer_new(writer_flags, configuration);
+  writer = log_writer_new(c.writer_flags, configuration);
 
   log_writer_set_options(writer, NULL, &opt, 0, 0, NULL, NULL);
   log_writer_set_queue(writer, queue);
-  log_writer_format_log(writer,msg,res);
-  if(strcmp(res->str,expected_value)!=0)
-    {
-      fprintf(stderr,"Testcase failed; result: %s, expected: %s\n",res->str,expected_value);
-      exit(1);
-    }
-  if (templ)
-    log_template_unref(templ);
-  log_pipe_unref((LogPipe *) writer);
-  log_msg_unref(msg);
-  log_queue_unref(queue);
-  g_string_free(res, TRUE);
+  log_writer_format_log(writer, msg, result_msg);
+  cr_assert_str_eq(result_msg->str, c.expected_value, "Expected: %s, actual: %s", c.expected_value, result_msg->str);
+
+  _tear_down(templ, writer, msg, queue, result_msg);
```

Comment:
If you had implemented this assert function as a [parameterized test case](http://criterion.readthedocs.io/en/master/parameterized.html#adding-parameterized-tests), you would have used Criterion's [teardown fixture](http://criterion.readthedocs.io/en/master/starter.html#fixtures) here.

Your solution is not redundant, maybe it is even easier to debug, so I'm OK with this.